File: pic2/pic2.f90
!-----------------------------------------------------------------------
! Skeleton 2D Electrostatic PIC code
! written by Viktor K. Decyk, UCLA
      program pic2
      use push2_h
      implicit none
! indx/indy = exponent which determines grid points in x/y direction:
! nx = 2**indx, ny = 2**indy.
      integer, parameter :: indx =   9, indy =   9
! npx/npy = number of electrons distributed in x/y direction.
      integer, parameter :: npx =  3072, npy =   3072
! ndim = number of velocity coordinates = 2
      integer, parameter :: ndim = 2
! tend = time at end of simulation, in units of plasma frequency.
! dt = time interval between successive calculations.
! qme = charge on electron, in units of e.
      real, parameter :: tend = 10.0, dt = 0.1, qme = -1.0
! vtx/vty = thermal velocity of electrons in x/y direction
! vx0/vy0 = drift velocity of electrons in x/y direction.
      real, parameter :: vtx = 1.0, vty = 1.0, vx0 = 0.0, vy0 = 0.0
! ax/ay = smoothed particle size in x/y direction
      real :: ax = .912871, ay = .912871
! idimp = number of particle coordinates = 4
! ipbc = particle boundary condition: 1 = periodic
! sortime = number of time steps between standard electron sorting
      integer :: idimp = 4, ipbc = 1, sortime = 50
! wke/we/wt = particle kinetic/electric field/total energy
      real :: wke = 0.0, we = 0.0, wt = 0.0
! declare scalars for standard code
      integer :: np, nx, ny, nxh, nyh, nxe, nye, nxeh, nxyh, nxhy
      integer :: ny1, ntime, nloop, isign
      real :: qbme, affp
!
! declare arrays for standard code:
! part, part2 = particle arrays
      real, dimension(:,:), pointer :: part, part2, tpart
! qe = electron charge density with guard cells
      real, dimension(:,:), pointer :: qe
! fxye = smoothed electric field with guard cells
      real, dimension(:,:,:), pointer :: fxye
! ffc = form factor array for poisson solver
      complex, dimension(:,:), pointer :: ffc
! mixup = bit reverse table for FFT
      integer, dimension(:), pointer :: mixup
! sct = sine/cosine table for FFT
      complex, dimension(:), pointer :: sct
! npicy = scratch array for reordering particles
      integer, dimension(:), pointer :: npicy
!
! declare and initialize timing data
      real :: time
      integer, dimension(4) :: itime
      real :: tdpost = 0.0, tguard = 0.0, tfft = 0.0, tfield = 0.0
      real :: tpush = 0.0, tsort = 0.0
      double precision :: dtime
!
! initialize scalars for standard code
! np = total number of particles in simulation
! nx/ny = number of grid points in x/y direction
      np = npx*npy; nx = 2**indx; ny = 2**indy; nxh = nx/2; nyh = ny/2
      nxe = nx + 2; nye = ny + 1; nxeh = nxe/2
      nxyh = max(nx,ny)/2; nxhy = max(nxh,ny); ny1 = ny + 1
! nloop = number of time steps in simulation
! ntime = current time step
      nloop = tend/dt + .0001; ntime = 0
      qbme = qme
      affp = real(nx*ny)/real(np)
!
! allocate data for standard code
      allocate(part(idimp,np))
      if (sortime > 0) allocate(part2(idimp,np))
      allocate(qe(nxe,nye),fxye(ndim,nxe,nye))
      allocate(ffc(nxh,nyh),mixup(nxhy),sct(nxyh))
      allocate(npicy(ny1))
!
! prepare fft tables
      call WFFT2RINIT(mixup,sct,indx,indy,nxhy,nxyh)
! calculate form factors
      isign = 0
      call POIS22(qe,fxye,isign,ffc,ax,ay,affp,we,nx,ny,nxeh,nye,nxh,nyh&
     &)
! initialize electrons
      call DISTR2(part,vtx,vty,vx0,vy0,npx,npy,idimp,np,nx,ny,ipbc)
!
! * * * start main iteration loop * * *
!
  500 if (nloop <= ntime) go to 2000
!     write (*,*) 'ntime = ', ntime
!
! deposit charge with standard procedure: updates qe
      call dtimer(dtime,itime,-1)
      qe = 0.0
      call GPOST2L(part,qe,qme,np,idimp,nxe,nye)
!     call GSPOST2L(part,qe,qme,np,idimp,nxe,nxe*nye)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tdpost = tdpost + time
!
! add guard cells with standard procedure: updates qe
      call dtimer(dtime,itime,-1)
      call AGUARD2L(qe,nx,ny,nxe,nye)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tguard = tguard + time
!
! transform charge to fourier space with standard procedure: updates qe
      call dtimer(dtime,itime,-1)
      isign = -1
      call WFFT2RX(qe,isign,mixup,sct,indx,indy,nxeh,nye,nxhy,nxyh)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tfft = tfft + time
!
! calculate force/charge in fourier space with standard procedure:
! updates fxye
      call dtimer(dtime,itime,-1)
      isign = -1
      call POIS22(qe,fxye,isign,ffc,ax,ay,affp,we,nx,ny,nxeh,nye,nxh,nyh&
     &)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tfield = tfield + time
!
! transform force to real space with standard procedure: updates fxye
      call dtimer(dtime,itime,-1)
      isign = 1
      call WFFT2R2(fxye,isign,mixup,sct,indx,indy,nxeh,nye,nxhy,nxyh)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tfft = tfft + time
!
! copy guard cells with standard procedure: updates fxye
      call dtimer(dtime,itime,-1)
      call CGUARD2L(fxye,nx,ny,nxe,nye)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tguard = tguard + time
!
! push particles with standard procedure: updates part, wke
      wke = 0.0
      call dtimer(dtime,itime,-1)
      call GPUSH2L(part,fxye,qbme,dt,wke,idimp,np,nx,ny,nxe,nye,ipbc)
!     call GSPUSH2L(part,fxye,qbme,dt,wke,idimp,np,nx,ny,nxe,nxe*nye,   &
!    &ipbc)
      call dtimer(dtime,itime,1)
      time = real(dtime)
      tpush = tpush + time
!
! sort particles by cell for standard procedure
      if (sortime > 0) then
         if (mod(ntime,sortime)==0) then
            call dtimer(dtime,itime,-1)
            call DSORTP2YL(part,part2,npicy,idimp,np,ny1)
! exchange pointers
            tpart => part
            part => part2
            part2 => tpart
            call dtimer(dtime,itime,1)
            time = real(dtime)
            tsort = tsort + time
         endif
      endif
!
      if (ntime==0) then
         write (*,*) 'Initial Field, Kinetic and Total Energies:'
         write (*,'(3e14.7)') we, wke, wke + we
      endif
      ntime = ntime + 1
      go to 500
 2000 continue
!
! * * * end main iteration loop * * *
!
      write (*,*) 'ntime = ', ntime
      write (*,*) 'Final Field, Kinetic and Total Energies:'
      write (*,'(3e14.7)') we, wke, wke + we
!
      write (*,*)
      write (*,*) 'deposit time = ', tdpost
      write (*,*) 'guard time = ', tguard
      write (*,*) 'solver time = ', tfield
      write (*,*) 'fft time = ', tfft
      write (*,*) 'push time = ', tpush
      write (*,*) 'sort time = ', tsort
      tfield = tfield + tguard + tfft
      write (*,*) 'total solver time = ', tfield
      time = tdpost + tpush + tsort
      write (*,*) 'total particle time = ', time
      wt = time + tfield
      write (*,*) 'total time = ', wt
      write (*,*)
!
      wt = 1.0e+09/(real(nloop)*real(np))
      write (*,*) 'Push Time (nsec) = ', tpush*wt
      write (*,*) 'Deposit Time (nsec) = ', tdpost*wt
      write (*,*) 'Sort Time (nsec) = ', tsort*wt
      write (*,*) 'Total Particle Time (nsec) = ', time*wt
!
      stop
      end program

File: serial/python_pic2/Pic2.py
#-------------------
#Skeleton 2D Electrostatic PIC code
# written by Thamine Dalichaouch, Viktor K. Decyk, Adam Tableman, and Qiyang Hu, UCLA
import numpy as np
import math
from push import *
import time
import matplotlib.pyplot as plt
int_type = np.int32
double_type = np.float64
float_type = np.float32
complex_type = np.complex64

def main():
	#----- Same format as other Skeleton Codes written by Viktor Decyk -----

	#number of cells in each direction-- nx = 2^indx
	indx = 8; indy=8
	#indx = 3; indy=3
	#number of particles in each direction
	#npx = 2024; npy = 2024
	npx = 500;npy = 500
	#ndim = number dimensions 
	ndim =2
	#tend = time at end of plasma simulation in units of plasma freq.
	#dt = time interval
	#qme = electron charge in units of |e|
	tend =500.0; dt = 0.1; qme = -1
	#vtx/vty are thermal velocities
	#vx0/vy0 are drift velocities
	vtx = 1.0; vty = 1.0; vx0 = 0; vy0 = 0;
	# ax/ay = smoothed particle size in x/y direction
	ax = .912871; ay = .912871
	# idimp = number of particle coordinates = 4
	# ipbc = particle boundary condition: 1 = periodic, 2 = reflecting
	# sortime = number of time steps between standard electron sorting
	idimp = 4; ipbc = 2; sortime = 50
	# wke/we/wt = particle kinetic/electric field/total energy
	wke = np.zeros((1),float_type)
	we = np.zeros((1),float_type)
	wt = np.zeros((1),float_type)

	# declare and initialize timing data
	itime = np.empty((4),int_type)
	tdpost = 0.0; tguard = 0.0; tfft = 0.0; tfield = 0.0
	tpush = 0.0; tsort = 0.0
	dtime = np.empty((1),double_type)

	# initialize scalars for standard code
	# nptot = total number of particles in simulation
	# nx/ny = number of grid points in x/y direction
	nptot = npx*npy; nx = int(math.pow(2,indx)); ny = int(math.pow(2,indy))

	# nyquist frequencies
	nxh = int(nx/2); nyh = max(1,int(ny/2))

	#guard cell definitions; do not change guard cells
	nxe = nx + 2; nye = ny + 1; 


	# nloop = number of time steps in simulation
	# ntime = current time step
	nloop = int(tend/dt +0.001); ntime = 0

	phi = np.empty((nloop,nxh,nyh),complex_type)
	phi.fill(0.0)
	#extra params
	qbme = qme 
	
	# inverse particles per cell normalizing factor
	affp = float(nx*ny)/nptot 

	# allocate data for code
	# part, part2 = particle arrays
	part = np.empty((idimp,nptot),float_type,'F')
	if(sortime>0):
		part2 = np.empty((idimp,nptot),float_type,'F')

	# electron charge density with guard cells
	qe = np.empty((nxe,nye),float_type,'F')
	qe2 = np.empty((nxe,nye),float_type,'F')

	# wavenumbers (ndim,kx,ky) for fourier analysis
	kvector = np.empty((ndim,nxe,nye),complex_type,'F')
	kvector.fill(0.0)

	#fft electron charge density
	qe_fft = np.empty((nxe,nye),complex_type,'F')

	# smoothed electric field with guard cells in fourier space
	fxye = np.empty((ndim,nxe,nye),complex_type,'F')

	# smoothed electric field with guard cells in real space
	fxyre = np.empty((ndim,nxe,nye),float_type,'F')
	
	# form factor array for poisson solver
	ffc = np.empty((nxe,nye),complex_type,'F')
	ffc.fill(0.0)

	#form factor array do this once
	form_factor(ffc,nx,ny,nxh,nyh,affp,ax,ay,kvector)

	#initialize particle velocities and 
	init_particles(part,vtx,vty,vx0,vy0,npx,npy,idimp,nptot,nx,ny,ipbc)
	
	#main loop
	for ntime in xrange(0,nloop):
		#zero matrices
		fxye.fill(0.0)
		qe.fill(0.0)
		fxyre.fill(0.0)
		qe_fft.fill(0.0)

		#deposit charge
		t = time.time()
		deposit2(part,qe,qme,nptot,idimp,nxe,nye)
		tdpost += time.time() - t
			
		
		#add guard cells
		t = time.time()
		guard_cells(qe,nx,ny,nxe,nye)
		tguard += time.time() - t

		
		#charge to fourier space!
		t = time.time()
		fourier_rho(qe_fft,qe,nx,ny)
		tfft += time.time() - t
		

		#calculate field in fourier space
		t = time.time()
		fourier_force(qe_fft,fxye,ffc,nx,ny,nxh,nyh,we,kvector)
		tfield += time.time()-t
		np.copyto(phi[ntime,:,:],fxye[0,:nxh,:nyh])	

		#force to real space, updates fxye
		t = time.time()
		real_force(fxyre,fxye,nx,ny,nxe,nye)
		tfft += time.time() -t
		
		t = time.time()
		guard_force(fxyre,nx,ny)
		tguard += time.time()-t


		#push particles updates part,wke
		wke[0] = 0.0
		t = time.time()
		push_particles(part,fxyre,qbme,nx,ny,dt,wke,ipbc,nptot,ndim)
		tpush += time.time() -t
		if(sortime > 0):
			if(ntime%sortime == 0):
				t = time.time()
				#sort by y axis
				part = part[:,np.argsort(part[1,:])]
				tsort = tsort + time.time() -t

		if (ntime<1):
			print "Initial Field, Kinetic and Total Energies:"
			print "%14.7e %14.7e %14.7e" % (we, wke, wke + we)
      		
      


	print "ntime = ", ntime
	print "Final Field, Kinetic and Total Energies:"
	print "%14.7e %14.7e %14.7e" % (we, wke, wke + we)

	print ""
	print "deposit time = ", tdpost
	print "guard time = ", tguard
	print "solver time = ", tfield
	print "fft time = ", tfft
	print "push time = ", tpush
	tfield = tfield + tguard + tfft
	print "sort time = ", tsort
	print "total solver time = ", tfield
	par_time = tdpost + tpush + tsort
	print "total particle time = ", par_time
	wt = par_time + tfield
	print "total time = ", wt
	print ""

	#diagnostics for code to view a plot of the disperion relation for Ex
	# default set to false
	print_diagnostics = True

	if(print_diagnostics):
		plt.figure(1,figsize=(12,6))
		freq_start = 0.7
		freq_end = 3
		fx_st = int(freq_start*dt*nloop/(2*np.pi))
		fx_end = int(freq_end*dt*nloop/(2*np.pi))
		print fx_st,fx_end
		freq_start = (2*np.pi/(dt*nloop))*fx_st
		freq_end = (2*np.pi/(dt*nloop))*fx_end
		plt.imshow(np.abs(np.fft.fft(phi[:,:,1],axis=0)[fx_st:fx_end,:]),origin ='lower',extent=[0,np.pi,freq_start,freq_end])
		#plt.axis([0,np.pi, freq_start,freq_end])
		plt.colorbar()
		plt.savefig('disp',dpi = 400,bbox_inches = 'tight')

	#plt.show()

if __name__ == "__main__":
	main()